CRM/Civioffice/DocumentRendererType/LocalUnoconv/PhpWordTemplateProcessor.php

```diff
@@ -13,124 +13,215 @@
+      $section = $phpWord->addSection();
+      // Note: addHtml() does not accept styles, so added HTML elements do not get applied any existing
+      // styles.
+      PhpWord\Shared\Html::addHtml($section, $rendered_token_message);
+      $elements = $section->getElements();
+      if ([] === $elements) {
+        // Note: If the paragraph had the macro as its only content, it
+        // will not be removed (i.e. leave an empty paragraph).
+        $this->setValue($macro_variable, '');
+      }
+      elseif (count($elements) === 1 && $elements[0] instanceof PhpWord\Element\Text) {
+        // ... either as plain text (if there is only a single Text
+        // element, which can't have style properties), ...
+        // Note: $rendered_token_message shouldn't be used directly
+        // because it may contain HTML entities.
+        $this->setValue($macro_variable, $elements[0]->getText());
+      }
+      else {
+        // ... or as HTML: Render all elements and insert in the text
+        // run or paragraph containing the macro.
+        $this->setElementsValue($macro_variable, $elements);
+      }
+    }
+    catch (Exception $exception) {
+      throw new CRM_Core_Exception(
+        E::ts('Error loading/writing PhpWord document: %1', [1 => $exception->getMessage()]),
+        $exception->getCode(),
+        [],
+        $exception
+      );
+    }
+    finally {
+      PhpWord\Settings::setOutputEscapingEnabled($outputEscapingEnabled);
+    }
+  }
+
+  /**
+   * Replaces a search string (macro) with a set of rendered elements, splitting
+   * surrounding texts, text runs or paragraphs before and after the macro,
+   * depending on the types of elements to insert.
+   *
+   * @param string $search
+   * @param \PhpOffice\PhpWord\Element\AbstractElement[] $elements
+   *
+   * @return void
+   * @throws \CRM_Core_Exception
+   */
+  public function setElementsValue(string $search, array $elements) {
+    $elements_data = '';
+    $hasParagraphs = FALSE;
+    foreach ($elements as $element) {
+      $elementName = substr(
+        get_class($element),
+        (int) strrpos(get_class($element), '\\') + 1
+      );
+      $objectClass = 'PhpOffice\\PhpWord\\Writer\\Word2007\\Element\\' . $elementName;
+
+      // For inline elements, do not create a new paragraph.
+      $withParagraph = 'PhpOffice\\PhpWord\\Writer\\Word2007\\Element\\Text' !== $objectClass;
+      $hasParagraphs = $hasParagraphs || $withParagraph;
+
+      $xmlWriter = new PhpWord\Shared\XMLWriter();
+      /** @var \PhpOffice\PhpWord\Writer\Word2007\Element\AbstractElement $elementWriter */
+      $elementWriter = new $objectClass($xmlWriter, $element, !$withParagraph);
+      $elementWriter->write();
+      $elements_data .= $xmlWriter->getData();
+    }
+    $blockType = $hasParagraphs ? 'w:p' : 'w:r';
+    $where = $this->findContainingXmlBlockForMacro($search, $blockType);
+    if (is_array($where)) {
```

Comment:
Is it a valid state if `$where` is not an array? If not, throw a `RuntimeException`?

Reply:
That is the case when `$search` (a PHPWord macro variable name) is not in the document. However, this is currently only being called for macro variables that exist in the document, as per https://github.com/systopia/de.systopia.civioffice/blob/73dc99c2b06310594099c4fbf73ae615b4b81524/CRM/Civioffice/DocumentRendererType/LocalUnoconv.php#L466

I think this method should just ignore non-existent tokens/macros in the document, as it does with this condition.